reclaim.c

```diff
@@ -158,7 +158,7 @@ GC_INLINE word *GC_clear_block(word *p, word sz, signed_word *count)
       p += 2;
     }
 # else
-    p++; /* Skip link field */
+    p += sizeof(word *); /* Skip link field */
```

Comment:
What type does `p` have at this point? Because I notice below that we say `*p++` which makes me unsure if this commit is complete.

Reply:
type has  `word *` at this  point.  This function is called in  [reclaim.c:L#200](https://github.com/capablevms/bdwgc/pull/5/commits/8aeb81e2bdcc074e8def623f78c632e4bf666a4b#L200)  and [reclaim.c:L263](https://github.com/capablevms/bdwgc/pull/5/commits/8aeb81e2bdcc074e8def623f78c632e4bf666a4b#L263). 

Immediately prior to these lines (in both instances) , the head of that  chunk is set to point to a backward pointer and the rest is being *zero'ed* out, i.e.  just skip the first "pointer" location. 

I did consider using a *capability* type and then using **NULL**.  What do you think ? It would be 

1. correct in *non*-CHERI  systems if  `typeof(NULL) == (long long)0` . However, in *CHERI* systems, this would mean the offset is *zero*, but the bounds are `0x0 - INTMAX`. The tag bit would be cleared however.  The question is whether this would constitute *clearing* the *rest* of the chunk. 
2. This could be *potentially* performance enhancing because  writing a **NULL** using a `void **` pointer type would be double the width of a `(long long)0` and the loop iterations would be half.

Reply:
> The tag bit would be cleared however. The question is whether this would constitute clearing the rest of the chunk.

Might be worth asking @jacobbramley. My guess is that this would depend on the particular CHERI architecture but I might be wrong!

Reply:
so long as the `GC_clear_block` function param `word* p` has permissive bounds (i.e. extending from initial value of `p` up to limit `q`) then the pointer arithmetic in this function all looks ok to me. Or are you worried about the cast from `word *` to `word` on  reclaim.c:L#162 ?

Reply:
> Or are you worried about the cast from word * to word on reclaim.c:L:162 ?

Exactly! That's why I'm wondering if `p += sizeof(word *)` is just a temporary hack but line 162 should also be adjusted (and, if it should, whether `p`'s type is wrong).

Reply:
this is a good shout, `p` used to advance one word (i.e. `p++`) so the linked list pointer was preserved. Now `p` should advance two words since the pointer is double-word sized. Since `p` is a pointer type, we *should not* be adding `sizeof(word *)` ... instead I think we should do:
`p += sizeof(word *)/sizeof(word);`

Reply:
PS I tested the code with this fixup and it still works! I think the previous version may have been skipping 16 words rather than 2 words, meaning some mark bits are never cleared

Reply:
> this is a good shout, `p` used to advance one word (i.e. `p++`) so the linked list pointer was preserved. Now `p` should advance two words since the link pointer is double-word sized. Since `p` is a pointer type, we _should not_ be adding `sizeof(word *)` ... instead I think we should do:
> `p += sizeof(word *)/sizeof(word);`

Yikes! Yes you are right Jeremy. Will fix.

Reply:
> PS I tested the code with this fixup and it still works! I think the previous version may have been skipping 16 words rather than 2 words, meaning some mark bits are never cleared

Yikes ! You are right Jeremy. Will fix